tools/run_functional_tests.py

```diff
@@ -0,0 +1,238 @@
+verboseprint = print if args.verbose else lambda *a, **k: None
+
+REPOSITORY_BIN = "_build/LEKA_V1_2_DEV/tests/functional/"
+EXTENSION_BIN = ".bin"
+
+LOOP_SLEEP_DELAY = float(args.loop_delay) if not args.fast else 0.1
+
+FILES_BIN = list()
+TESTS = list()
+
+FAILED_STR = Fore.RED + "❌FAILED" + Style.RESET_ALL
+SUCCESS_STR = Fore.GREEN + "✅SUCCESS" + Style.RESET_ALL
+
+
+class Test:
+    saving_repository = "./fs/usr/test/"
+    saving_extension_txt = ".txt"
+    saving_default_filename = "file_saving"
+    saving_filepath = saving_repository + \
+        saving_default_filename + saving_extension_txt
+
+    def __init__(self, path):
+        self.path = path
+        self.failures = list()
+
+    def generate_saving_file(self):
+        def define_path():
+            base = os.path.basename(self.path)
+            stem, ext = os.path.splitext(base)
+            ret = self.saving_repository + stem + self.saving_extension_txt
+            return ret
+
+        def create_file(path):
+            try:
+                f = open(path, "w")
+            except OSError as e:
+                print("Could not create or open file: " + path)
+                print("Error: " + e)
+                sys.exit()
+            f.close()
+
+        self.saving_filepath = define_path()
+        create_file(self.saving_filepath)
+
+    def edit_saving_file(self, data):
+        file = open(self.saving_filepath, "a")
+        file.write(data.strip() + '\n')
+        file.close()
+
+    def print_saving_file(self, color=Fore.RED):
+
+        file = open(self.saving_filepath, "r")
+        data = file.read()
+        file.close()
+        verboseprint("\n"+self.saving_filepath)
+        verboseprint(color + data + Style.RESET_ALL)
+
+    @staticmethod
+    def delete_saving_files(self):
+        files = glob.glob(self.saving_repository)
+        for f in files:
+            os.remove(f)
+
+    def run(self):
+        def reboot_device():
+            com.send_break()
+            REBOOT_SLEEP_DELAY = 3
+            sleep(REBOOT_SLEEP_DELAY)
+
+        def flash():
+            verboseprint("Flashing..." + self.path)
+            CMD_FLASH = "openocd -f interface/stlink.cfg -c 'transport select hla_swd' -f target/stm32f7x.cfg -c 'program " + \
+                self.path+" 0x08000000' -c exit"
+            os.system(CMD_FLASH)
+            sleep(1)
+            CMD_RESET = "openocd -f interface/stlink.cfg -c 'transport select hla_swd' -f target/stm32f7x.cfg -c init -c 'reset run' -c exit"
+            os.system(CMD_RESET)
+
+        def read_output_serial():
+            return com.readline().decode("utf-8")
+
+        def wait_for_response():
+            data = ''
+            no_response_counter = 0
+
+            while (no_response_counter <= MAX_GET_LINE_RETRIES):
+                sleep(.005)
+                data = read_output_serial()
+                if (data):
+                    return data
+                no_response_counter += 1
+
+            return None
```

Comment:
I find nested function definitions hard to read. 

Those are very related, they could be moved to their own class